CMakeLists.txt

```diff
@@ -0,0 +1,225 @@
+cmake_minimum_required(VERSION 3.14)
+project(mold VERSION 1.3.0)
+
+include(CMakeDependentOption)
+include(GNUInstallDirs)
+
+# Mold build options
+cmake_dependent_option(MOLD_USE_MIMALLOC "Use mimalloc" ON
+    "NOT CMAKE_SYSTEM_NAME MATCHES Darwin; NOT Android" OFF)
+option(MOLD_USE_SYSTEM_MIMALLOC "Use system mimalloc")
+option(MOLD_USE_SYSTEM_TBB "Use system TBB")
+option(MOLD_USE_LTO "Use Link Time Optimization")
+option(CMAKE_INSTALL_LOCAL_ONLY ON)
+
+# Default to `Release` build type,
+# to be consistent with the current Makefiles
+if(NOT CMAKE_BUILD_TYPE)
+    set(CMAKE_BUILD_TYPE Release CACHE STRING "" FORCE)
+endif()
+
+# Setup mimalloc
+if(MOLD_USE_MIMALLOC)
+    if(MOLD_USE_SYSTEM_MIMALLOC)
+        find_library(MIMALLOC_LIBRARIES mimalloc REQUIRED)
+        add_library(mimalloc INTERFACE)
+        target_link_libraries(mimalloc INTERFACE ${MIMALLOC_LIBRARIES})
+        target_compile_definitions(mimalloc INTERFACE USE_SYSTEM_MIMALLOC)
+    else()
+        function(mold_add_mimalloc)
+            string(APPEND CMAKE_C_FLAGS -DMI_USE_ENVIRON=0)
+            option(MI_BUILD_TESTS "Build test executables" OFF)
+            add_subdirectory(third-party/mimalloc EXCLUDE_FROM_ALL)
+        endfunction()
+
+        mold_add_mimalloc()
+    endif()
+
+    add_library(mimalloc::mimalloc ALIAS mimalloc)
+endif()
+
+# Setup TBB
+if(MOLD_USE_SYSTEM_TBB)
+    find_package(TBB REQUIRED)
+else()
+    function(mold_add_tbb)
+        set(CMAKE_BUILD_TYPE RelWithDebInfo)
+        set(BUILD_SHARED_LIBS OFF)
+        option(TBB_TEST "Enable testing" OFF)
+        string(APPEND CMAKE_CXX_FLAGS -D__TBB_DYNAMIC_LOAD_ENABLED=0)
+        option(TBB_STRICT "Treat compiler warnings as errors" OFF)
+        add_subdirectory(third-party/tbb EXCLUDE_FROM_ALL)
+    endfunction()
+
+    mold_add_tbb()
+endif()
+
+if(MOLD_USE_LTO)
+    include(CheckIPOSupported)
+    check_ipo_supported()
+endif()
+
+include(CheckPIESupported)
+check_pie_supported()
+
+add_library(mold-wrapper SHARED)
+set_target_properties(mold-wrapper PROPERTIES
+    POSITION_INDEPENDENT_CODE ON
+    PREFIX "" # Remove the default `lib` prefix
+    INTERPROCEDURAL_OPTIMIZATION ${MOLD_USE_LTO})
+target_link_libraries(mold-wrapper PRIVATE ${CMAKE_DL_LIBS})
+target_sources(mold-wrapper PRIVATE elf/mold-wrapper.c)
+
+add_executable(mold)
+target_compile_features(mold PRIVATE cxx_std_20)
+set_target_properties(mold PROPERTIES
+    CXX_EXTENSIONS OFF
+    INTERPROCEDURAL_OPTIMIZATION ${MOLD_USE_LTO})
+target_include_directories(mold PRIVATE third-party)
+
+if(NOT MSVC)
+    target_compile_options(mold PRIVATE
+        $<$<PLATFORM_ID:Android>:-Wno-c++11-narrowing>
+        -fno-exceptions
+        -fno-unwind-tables
+        -fno-asynchronous-unwind-tables)
+
+    find_package(ZLIB REQUIRED)
+    find_library(M_LIBRARIES m REQUIRED)
```

Comment:
Just because libm exists, doesn't mean it's necessary to be used. The check example determines exactly whether it's needed or not and you can act accordingly.

Reply:
Yes, you are right. Another question, is there any better alternative to `target_link_libraries` when it comes to linking a library like libm, which we know is not a CMake target. I never know if `target_link_libraries` is really the best solution in that case, as it might prefer a CMake target with the same name instead...

Reply:
`target_link_libraries` **is** the way you link to targets or pass library names to the compiler. You don't have to worry about the names here, open source developers are generally pretty good with not stepping on others' toes with name conflicts, especially when it comes to such fundamental ones like libm.

Reply:
Okay, Thanks.